apsuite/optics_analysis/orbit_correction.py

```diff
@@ -0,0 +1,164 @@
+    """."""
+
+    MAX_HKICK = 300e-6  # urad
+    MAX_VKICK = 300e-6  # urad
+    CORR_STATUS = _namedtuple('CorrStatus', ['Fail', 'Sucess'])(0, 1)
+
+    def __init__(self, model, acc, dim='4d'):
+        """."""
+        self.model = model
+        self.acc = acc
+        self.dim = dim
+        self.respm = OrbRespmat(model=self.model, acc=self.acc, dim=self.dim)
+        self.bpm_idx = self.respm.fam_data['BPM']['index']
+        self.ch_idx = self.respm.fam_data['CH']['index']
+        self.cv_idx = self.respm.fam_data['CV']['index']
+        self.corr_idx = np.vstack((self.ch_idx, self.cv_idx))
+
+    @property
+    def _nbpm(self):
+        return len(self.bpm_idx)
+
+    @property
+    def _nch(self):
+        return len(self.ch_idx)
+
+    @property
+    def _ncv(self):
+        return len(self.cv_idx)
+
+    @property
+    def _ncorr(self):
+        return self._nch + self._ncv
+
+    def calc_jacobian_matrix(self, model=None):
+        """."""
+        if model is None:
+            model = self.model
+        self.respm.model = model
+        return self.respm.get_respm()
+
+    def _get_orbit_residue(self, model):
+        if self.dim == '4d':
+            cod = pyaccel.tracking.find_orbit4(model, indices='open')
+        elif self.dim == '6d':
+            cod = pyaccel.tracking.find_orbit6(model, indices='open')
+        codx = cod[0, self.bpm_idx].flatten()
+        cody = cod[2, self.bpm_idx].flatten()
+        res = np.hstack((codx, cody))
+        return res
+
+    def get_kicks(self, model=None, chidx=None, cvidx=None):
+        """Return skew quadrupoles strengths."""
+        if model is None:
+            model = self.model
+        if chidx is None:
+            chidx = self.ch_idx
+        if cvidx is None:
+            cvidx = self.cv_idx
+        hkick = []
+        for mag in chidx:
+            hkick_seg = []
+            for seg in mag:
+                hkick_seg.append(model[seg].hkick_polynom)
+            hkick.append(hkick_seg)
+        vkick = []
+        for mag in cvidx:
+            vkick_seg = []
+            for seg in mag:
+                vkick_seg.append(model[seg].vkick_polynom)
+            vkick.append(vkick_seg)
+        hkick = np.array(hkick)
+        vkick = np.array(vkick)
+        return np.vstack((hkick, vkick))
+
+    def set_kicks(self, model=None, corridx=None, kicks=None):
+        """Set skew quadrupoles strengths in the model."""
+        if model is None:
+            model = self.model
+        if corridx is None:
+            corridx = self.corr_idx
+        if kicks is None:
+            raise Exception('Missing Kicks values')
+        newmod = _dcopy(model)
+        for idx_mag, mag in enumerate(corridx):
+            for idx_seg, seg in enumerate(mag):
+                kick = kicks[idx_mag][idx_seg]
+                if idx_mag < self._nch:
+                    if np.abs(kick) > OrbitCorr.MAX_HKICK:
+                        kick = np.sign(kick) * OrbitCorr.MAX_HKICK
```

Comment:
I think this won't work well. When you reach the MAX_KICK you have three options: 
 - scale all kicks in such a way that the maximum value is equal to MAX_KICK;
 - give up, raise an error or whatever, so that the user knows about this and remove by hand that corrector from the correction;
 - remove the corrector from the matrix, and calculate the inverse once again.

I think the easiest way is to give up and let the user handle this problem.

Reply:
Ok, now this case is handled by the new `_check_kicks` method.

Reply:
see comment below.